Source/drlg_l1.cpp

```diff
@@ -2168,6 +2168,130 @@ void CryptPatternGroup7(int rndper)
 	PlaceMiniSetRandom(CryptPattern8, rndper);
 }
 
+bool PlaceCathedralMiniSets(lvl_entry entry)
+{
+	bool doneflag = false;
+
+	if (Quests[Q_PWATER].IsAvailable()) {
+		if (entry == ENTRY_MAIN) {
+			if (PlaceMiniSet(PWATERIN, 1, 1, 0, 0, true, -1) < 0)
+				doneflag = true;
+		} else {
+			if (PlaceMiniSet(PWATERIN, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+			ViewPosition.y--;
+		}
+	}
+	if (Quests[Q_LTBANNER].IsAvailable()) {
+		if (entry == ENTRY_MAIN) {
+			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, true, -1) < 0)
+				return true;
+		} else {
+			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+			if (entry == ENTRY_PREV) {
+				ViewPosition = Point { 20, 28 } + Displacement { setpc_x, setpc_y } * 2;
+			} else {
+				ViewPosition.y--;
+			}
+		}
+	} else if (entry == ENTRY_MAIN) {
+		if (Players[MyPlayerId].pOriginalCathedral) {
+			if (PlaceMiniSet(L5STAIRSUP, 1, 1, 0, 0, true, -1) < 0)
+				return true;
+			if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0)
+				return true;
+		} else {
+			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, true, -1) < 0)
+				doneflag = true;
+			if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;

```

Comment:
When `PlaceMiniSet` is called with `tmax - tmin == 0` it calls the global rng exactly twice. This could use an early return like:
```suggestion
			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, true, -1) < 0) {
				// Skip trying to place the down stairs since we've already failed to generate the level, but we still need to advance the RNG state twice to match vanilla dungeon generation.
				AdvanceRndSeed();
				AdvanceRndSeed();
				return true;
			}
			if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0)
				return true;
```